nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java

```diff
@@ -0,0 +1,230 @@
+        @WritesAttribute(attribute = "s3.length", description = "The size of the object in bytes"),
+        @WritesAttribute(attribute = "s3.storageClass", description = "The storage class of the object"),})
+@SeeAlso({FetchS3Object.class, PutS3Object.class, DeleteS3Object.class})
+public class ListS3 extends AbstractS3Processor {
+
+    public static final PropertyDescriptor DELIMITER = new PropertyDescriptor.Builder()
+            .name("Delimiter")
+            .expressionLanguageSupported(false)
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .description("The string used to delimit directories within the bucket. Please consult the AWS documentation " +
+                    "for the correct use of this field.")
+            .build();
+
+    public static final PropertyDescriptor PREFIX = new PropertyDescriptor.Builder()
+            .name("Prefix")
+            .expressionLanguageSupported(false)
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .description("The prefix used to filter the object list. In most cases, it should end with a forward slash ('/').")
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(BUCKET, KEY, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE, TIMEOUT, SSL_CONTEXT_SERVICE, ENDPOINT_OVERRIDE,
+                    DELIMITER, PREFIX));
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(
+            new HashSet<>(Collections.singletonList(REL_SUCCESS)));
+
+    public static final String CURRENT_TIMESTAMP = "currentTimestamp";
+    public static final String CURRENT_KEY_PREFIX = "key-";
+
+    // State tracking
+    private long currentTimestamp = 0L;
+    private Set<String> currentKeys;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    private Set<String> extractKeys(final StateMap stateMap) {
+        Set<String> keys = new HashSet<>();
+        for (Map.Entry<String, String>  entry : stateMap.toMap().entrySet()) {
+            if (entry.getKey().startsWith(CURRENT_KEY_PREFIX)) {
+                keys.add(entry.getValue());
+            }
+        }
+        return keys;
+    }
+
+    private void restoreState(final ProcessContext context) throws IOException {
+        final StateMap stateMap = context.getStateManager().getState(Scope.CLUSTER);
+        if (stateMap.getVersion() == -1L) {
+            currentTimestamp = 0L;
+            currentKeys = new HashSet<>();
+        } else {
+            currentTimestamp = Long.parseLong(stateMap.get(CURRENT_TIMESTAMP));
```

Comment:
The ListS3 processor throws exceptions on restart if state was cleared through the UI while the processor was stopped.

2016-02-22 14:00:05,586 ERROR [Timer-Driven Process Thread-3] org.apache.nifi.processors.aws.s3.ListS3 ListS3[id=5763acff-1b63-461a-bb72-fd12bd7ac603] ListS3[id=5763acff-1b63-461a-bb72-fd12bd7ac603] failed to process due to java.lang.NumberFormatException: null; rolling back session: java.lang.NumberFormatException: null
2016-02-22 14:00:05,637 ERROR [Timer-Driven Process Thread-3] org.apache.nifi.processors.aws.s3.ListS3 
java.lang.NumberFormatException: null
	at java.lang.Long.parseLong(Long.java:404) ~[na:1.7.0_80]
	at java.lang.Long.parseLong(Long.java:483) ~[na:1.7.0_80]
	at org.apache.nifi.processors.aws.s3.ListS3.restoreState(ListS3.java:133) ~[na:na]
	at org.apache.nifi.processors.aws.s3.ListS3.onTrigger(ListS3.java:155) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:49) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:124) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [na:1.7.0_80]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304) [na:1.7.0_80]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178) [na:1.7.0_80]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_80]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_80]